internal/job/restore/restore.go

```diff
@@ -158,31 +160,46 @@ func (r *Restore) loopCopyChunk(privateData *restorePrivateData, rawImageSize ui
 	return nil
 }
 
-func (r *Restore) doRestoreDiffPhase(privateData *restorePrivateData, diffs []*job.BackupMetadataEntry) error {
+func (r *Restore) doRestoreDiffPhase(privateData *restorePrivateData, metadata *job.BackupMetadata) error {
 	if privateData.Phase != RestoreDiff {
 		return nil
 	}
 
-	for _, diff := range diffs {
-		err := r.loopApplyDiff(privateData, diff)
+	// TODO: Multiple diffs are not currently supported.
+	// This logic only works correctly if metadata.Diff contains at most one diff.
+	source := metadata.Raw
+	for _, target := range metadata.Diff {
+		err := r.loopApplyDiff(privateData, source, target)
 		if err != nil {
 			return fmt.Errorf("failed to apply diffs: %w", err)
 		}
-		if diff.SnapID == r.targetSnapshotID {
+		if target.SnapID == r.targetSnapshotID {
 			break
 		}
+		source = target
 	}
 
 	privateData.Phase = Completed
 	return setRestorePrivateData(r.repo, r.actionUID, privateData)
 }
 
-func (r *Restore) loopApplyDiff(privateData *restorePrivateData, diff *job.BackupMetadataEntry) error {
-	partCount := int(math.Ceil(float64(diff.SnapSize) / float64(diff.PartSize)))
+func (r *Restore) loopApplyDiff(
+	privateData *restorePrivateData,
+	source *job.BackupMetadataEntry,
+	target *job.BackupMetadataEntry) error {
+	partCount := int(math.Ceil(float64(target.SnapSize) / float64(target.PartSize)))
 	for i := privateData.NextDiffPart; i < partCount; i++ {
-		if err := r.restoreVol.ApplyDiff(
-			r.nodeLocalVolumeRepo.GetDiffPartPath(diff.SnapID, i),
-		); err != nil {
+		sourceSnapshotName := source.SnapName
+		if i != 0 {
+			sourceSnapshotName = fmt.Sprintf("%s-offset-%d", target.SnapName, uint64(i)*target.PartSize)
+		}
+		targetSnapshotName := target.SnapName
+		if i != partCount-1 {
+			targetSnapshotName = fmt.Sprintf("%s-offset-%d", target.SnapName, uint64(i+1)*target.PartSize)
+		}
```

Comment:
[internal/job/backup/backup.go](https://github.com/cybozu-go/fin/blob/86ab9f59589cd55de321b771c48f07cbe32d8191/internal/job/backup/backup.go#L293-L300) already has similar code, so we may be able to make a new utility function that can be used from both sides with respect to DRY principle.